qldpc/abstract.py

```diff
@@ -531,3 +607,122 @@ def lift(member: int) -> IntegerArray:
+            gen_x[0, 1] = 1
+        else:
+            gen_x[0, 0] = self.field.primitive_element
+            gen_x[1, 1] = self.field.primitive_element**-1
+            gen_w[0, 0] = -1 * self.field(1)
+        return gen_x, gen_w
+
+    def target_space(self) -> list[bytes]:
+        """All members of the target space that this group acts on."""
+        vectors = itertools.product(self.field.elements, repeat=self._dimension)
+        next(vectors)  # skip the all-0 element
+        return [self.field(vec).tobytes() for vec in vectors]
+
+    @classmethod
+    def iter_mats(cls, dimension: int, field: int | None = None) -> Iterator[galois.FieldArray]:
+        """Iterate over all elements of SL(dimension, field)."""
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        for entries in itertools.product(base_field.elements, repeat=dimension**2):
+            vec = base_field(entries)
+            mat = vec.reshape(dimension, dimension)
+            if np.linalg.det(mat) == 1:
+                yield mat  # type:ignore[misc]
+
+
+class ProjectiveSpecialLinearGroup(Group):
+    """Projective variant of the special linear group (PSL)."""
+
+    _dimension: int
+
+    def __init__(self, dimension: int, field: int | None = None) -> None:
+        self._dimension = dimension
+        self._field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        if self.field.order == 2:
+            super().__init__(SpecialLinearGroup(dimension, 2))
+        elif dimension == 2:
+            generator_mats = self.get_generator_mats()
+            target_space = self.target_space()
+            group = Group.from_generating_mats(generator_mats, target_space, field)
+            super().__init__(group)
+        else:
+            raise ValueError(
+                "Projective special linear groups with both dimension and field greater than 2 are"
+                " not yet supported."
+            )
+
+    @property
+    def dimension(self) -> int:
+        """Dimension of the elements of this group."""
+        return self._dimension
+
+    def get_generator_mats(self) -> tuple[galois.FieldArray, ...]:
+        """Expanding generator matrices for this group.
+
+        This construction is based on https://arxiv.org/abs/1807.03879.
+        """
+        minus_one = -self.field(1)
+        A = self.field([[1, 1], [0, 1]])
+        B = self.field([[1, minus_one], [0, 1]])
+        C = self.field([[1, 0], [1, 1]])
+        D = self.field([[1, 0], [minus_one, 1]])
+        return A, B, C, D
+
+    def target_space(self) -> list[bytes]:
+        """All members of the target space that this group acts on.
+
+        Since this group acts on a projective space, scalar multiples of target vectors are
+        identified with each other.
+        """
+        return [
+            self.field(vec).tobytes()
+            for vec in itertools.product(self.field.elements, repeat=self.dimension)
+            if vec[(self.field(vec) != 0).argmax()] == 1
+        ]
```

Comment:
This construction makes sense to me, but it's not obvious to me that it's compatible with the current implementation of `Groups.from_generating_mats`.  For convenience, we operationally identify the target space of PSL(d,q) with the set of vectors whose first element is 1.  However, I'm not sure that this space is actually stabilized by the matrices that we use to represent PSL(d,q), which means that [this line](https://github.com/Infleqtion/qLDPC/blob/2c2b21194eaf19812c69ba6c840eaea4db5b5f27/qldpc/abstract.py#L260) would fail to find a `next_index`.

Thoughts?

Reply:
I tested and the construction is indeed broken.  I'll remove this construction of PSL and do something else for now.

Reply:
![image](https://github.com/Infleqtion/qLDPC/assets/567092/45755657-dfd5-4101-bd63-73577aa63a3c)